### ISSUE TITLE
Can't make Makefile.win

### ISSUE THREAD

**Issue by Sonny-Dreams** [ppcasm/gsuploader#1]
Hello,

I have been trying to compile gsuploader on windows. Is using nmake in visual studio the easiest way? I keep getting this error

"NMAKE : fatal error U1073: don't know how to make 'all-before'
Stop."

**Comment by ppcasm** [comment]
I used dev-cpp to compile it on windows. It was an older version and I
haven't compiled this version on Windows in a while. Have you tried the usb
version? I also wrote a very beta utility for using the official gameshark
tools with the usb dongle but you have to have a moschip based one for the
time being. If that's not an option then just try compiling with an older
devcpp version. I think is was devcpp version 4.9.9.2 The parallel port
version is a clusterfuck and you also have to be aware of whether your
gameshark has the internal components for it. Version 3.3 was hit or miss
if it had them, every version 3.2 gameshark I've seen has the internal
components for it, in fact I've recently been buying v3.2 gamesharks on
eBay and socketing the eeproms so that I can hopefully eventually get
around to writing a custom gameshark firmware that isn't a pile of shit.
Let me know if you need help setting it up, or alternatively I could
compile a Windows version for you when i get some time. I'm not really sure
why I didn't provide a binary.

**Comment by Sonny-Dreams** [comment]
Wow thanks for responding so quickly! I installed an older version of devcpp but it can't compile on my computer (I'd probably need a virtual machine). 

Yeah I already learned the hard way about the V3,3 parallel port. I read somewhere that all 3.3's in a shiny case have the right parts. When I looked inside mine, it was missing everything. I then got a V3.2 and it seems like it will work. My main problem is trying to get a parallel port. I cannot find that moschip dongle on ebay or amazon. I think it might actually be easier for me to find an old computer from a salvage store.

If you could compile the windows binary that would be amazing

**Comment by ppcasm** [comment]
https://www.amazon.com/gp/product/B000V6Y7UW/ref=oh_aui_detailpage_o04_s00?ie=UTF8&psc=1

**Comment by ppcasm** [comment]
Did you get the $8 adapter? If you can't get a hold of it let me know. I
don't think you want to deal with the headache of doing anything with a
mainboard parallel port unless you legit are running windows 98 as the main
OS.

**Comment by Sonny-Dreams** [comment]
Yeah I ordered it, but it isn't in yet. Yeah I don't wanna deal with finding an old computer that has a parallel port haha. I'll try it out as soon as it gets in

**Comment by ppcasm** [comment]
Let me know when it does as the stuff I've wrote (mappar in particular) is
largely untested, but something I'm definitely interested in working on and
perfecting for not only n64 gameshark offical tool use, but the whole
parallel port thing in general. There really needs to be a solution for the
parallel port problem. Not sure if you checked this out or not:
http://ppcasm.blogspot.com/2014/09/the-parallel-port-problem.html

I've tested this on a few devices (mainly eeprom programmers and stuff) and
it was stupid slow for those, but it worked. It works fairly quickly with
gsuploader USB version because we patched out the reply stuff so it just
kind of bulk sends the code to N64 and doesn't have to do a bunch of USB
transactions. I've only recently freed up a little time where I might be
able to start working on these things again.

**Comment by ppcasm** [comment]
The other cool part about the USB thing is that it also works with a
raspberry pi and such, and I've even installed an n64 toolchain on a
raspberry pi and it allows you to program n64 through WIFI haha

**Comment by Sonny-Dreams** [comment]
Ok I got my cable in and I successfully compiled the code on linux. However, I keep getting "failed to locate/open device". 

My parallel port shows up as "Bus 001 Device 007: ID 9710:7705 MosChip Semiconductor MC7705 Parallel Port Adapter". I tried using MicroMachines with codegeneration on but I keep getting the same error. For some reason non of my other games will boot up with my game shark. I somehow lost my Mario 64 copy

Do you think it's a problem with it finding the adapter? Or do I just need to try another game

**Comment by ppcasm** [comment]
With the default keycode (Don't change this unless you know for sure you
have a supported game) it'll only boot 6102 CIC games (such as mario64,
007, waverace, etc) . You have to install libusb1.0-dev example: sudo
apt-get install libusb-1.0-0-dev

**Comment by ppcasm** [comment]
if the game booted then it should be okay. It's the right adapter as well.
Did you have to mod it to make it fit? I've had to always break off the
post connectors on the port. In any case, it's likely because of
libusb1.0-dev not being installed

**Comment by Sonny-Dreams** [comment]
Ok I just got Starfox to startup. I had to install libusb1.0-dev in order to get the "make" command to work. 

I actually just took the plastic case off of the gameshark so that the post connectors didn't stop the connector from touching

**Comment by ppcasm** [comment]
It still isn't recognizing? I wonder if there's another driver that needs to be blacklisted or something. You definitely have the right adapter and hardware

**Comment by ppcasm** [comment]
Yeah, I usually take a pair of channel locks and bend them outward until they break so they have a downward recess

**Comment by ppcasm** [comment]
Wait... Are you running as sudo?

**Comment by Sonny-Dreams** [comment]
Yep! Just realized I needed to run as sudo. Oh man. It interupted starfox and tried to load but then I got "async transfer failed: LIBUSB_TRANSFER_TIMED_OUT"

I'm just gonna test it a few more times. I think I am really close

**Comment by ppcasm** [comment]
Yeah, it's tricky. You have to pay attention to where you upload code, and where things are compiled to run at entry and what not. I can't remember if a crash gives that error or not, but the tricky part is trying to find a good area to upload to that doesn't corrupt the game while it's running, or other useful things. It's a balancing act that would probably be better suited to have another loader reside in the very end of expansion ram that's unlikely to be used under normal circumstances and have that do the loading. But for now we rely on the gameshark built in funcs.

**Comment by ppcasm** [comment]
The best idea is to test with neon64 first

**Comment by Sonny-Dreams** [comment]
Yeah I have been testing with neon64. Ill try again tomorrow. 

![image](https://user-images.githubusercontent.com/14127899/48668418-8de1c400-eaba-11e8-9649-fb7e07534816.png)

This is the output from two different tries

**Comment by ppcasm** [comment]
Did you try micromachines game? I'll have to see what it does because I can't remember if comms getting corrupted will cause this error, or if it's something in libusb/somewhere else. I'm thinking it's corrupting the game/comms though.

**Comment by ppcasm** [comment]
Did you get it working?

**Comment by Sonny-Dreams** [comment]
Sorry had more time today. It crashes around the same point in Starfox, Pilotwings, and Micro Machines. I think its something in libusb

**Comment by Sonny-Dreams** [comment]
Actually I think it may be because I am running linux inside a VM and the timings might be messed up. I will play with some of the VM settings to see if I can change anything

**Comment by Sonny-Dreams** [comment]
I got it to work! I added an extension to Virtualbox to support usb3.0 and I started the upload while in the middle of a game in Pilotwings

**Comment by ppcasm** [comment]
Interesting. Did you try with other games? I'm interested in times when it
doesn't work in case I need to do some relocation of the preloader. I'm
thinking of actually just writing a raw loader and having it in a safe
place. I think I have the N64 side of the loader done... somewhere. I'd
just need to do the PC side, and have it run from exp ram that's likely
never, or rarely used.
